mysys/my_largepage.c

```diff
@@ -79,88 +69,353 @@ void my_large_free(uchar* ptr)
     my_large_malloc_int(), i.e. my_malloc_lock() was used so we should free it
     with my_free_lock()
   */
-  if (!my_use_large_pages || !my_large_page_size || !my_large_free_int(ptr))
+  if (!my_large_free_int(ptr, size))
     my_free_lock(ptr);
+  else
+    MEM_UNDEFINED(ptr, size);
```

Comment:
For every byte of memory in a program ASAN remembers whether it's poisoned or not. It has some huge array or something similar. On allocating functions like `malloc()` it unpoisons memory, on freeing function like `free()` it poisons memory. But for functions like `munmap()` it does nothing. If memory was poisoned it will remains so. Then OS may reuse that memory for some TLS or stack variable and for ASAN it will be automatically poisoned. And we'll have hard to debug false positives like in https://jira.mariadb.org/browse/MDEV-21239

In short, someone should unpoison potentially poisoned memory right before or right after (it doesn't matter when) `munmap()` or `VirtualFree()`.

Reply:
Ok. Thanks. Reused your words in the comments.